### DISCUSSION TITLE
Add debuggable steps

### DISCUSSION THREAD

**Discussion by Eriner** [woodpecker-ci/woodpecker#2290] [Request a Feature]
### Clear and concise description of the problem

Sometimes builds exhibit problems in CI that are not reproducible on a developer machine. Often this has to do with networking, environment, or other oversights. Often, if at a shell, a developer could easily solve these problems. Instead, they must push many "printfs".

### Suggested solution

Add the ability for developers to connect to a shell via ssh at a step.

### Alternative

Status quo (same with GitHub Actions, Drone, and others), developers must "printf" or painfully modify their build to minimize build times between steps for debugging speed.

### Additional context

similar: #244

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by 6543** [comment]
I think we should solve this in a generic way:
would require to add a new function  "attach" to pipeline backend interface

and implement it ...

**Comment by 6543** [comment]
questioins to solve:
 - how do you specify a step to be debugged
 - at whitch point you should be able to attach
 - make sure steps are not closed right away ... if?

**Comment by johanvdw** [comment]
Drone allows starting a tmate debug session. 
https://community.harness.io/t/feature-preview-debug-mode/11474/1
Also for github actions such a plugin exists: 
https://github.com/marketplace/actions/debugging-with-tmate-and-zsh

I do think restarting a pipeline with a tmate terminal in a specific place may be a nice solution.

**Reply by wailashi** [reply]
With Concourse CI you can open a shell in any currently or recently running build container using their command line tool.
Especially nice that it doesn't require any changes to the pipeline.

**Comment by johanvdw** [comment]
FYI: I've recently added this in a pipeline where I would like to debug, quite nice if you allow outgoing ssh. You can also host your own uptermd server

```
  commands:
  - wget -q https://github.com/owenthereal/upterm/releases/download/v0.22.0/upterm_linux_amd64.tar.gz
  - tar xvzf upterm*
  - ssh-keygen -t ed25519 -N "" -f ~/.ssh/id_ed25519
  - ./upterm host --accept  --skip-host-key-check  bash
```
